### PR TITLE
Multiple apps

### PR DESCRIPTION
**Pull request by amstone326**

### CODE REVIEW

**Comment by czue** [comment]
:scream_cat:

**Comment by orangejenny** [comment]
x-requested with https://github.com/dimagi/commcare-odk/pull/131

File: backend/src/org/commcare/suite/model/Profile.java [hunk 1]
```diff
@@ -84,6 +88,21 @@ public Profile(int version, String authRef) {
...
+	 */
+	public String getUniqueId() {
+		return this.uniqueId;
+	}
+	
+	/*
+	 * Return whether this Profile was generated from an old version of profile.ccpr
+	 * (i.e. does not have a uniqueId or displayName)
+	 */
+	public boolean fromOld() {
```
Comment (orangejenny) [inline backend/src/org/commcare/suite/model/Profile.java]: Good that this is commented, but it's pretty vague. How about returning !fromOld and calling it multipleAppsEnabled?

File: backend/src/org/commcare/xml/ProfileParser.java [hunk 2]
```diff
@@ -48,46 +47,57 @@ public Profile parse() throws InvalidStructureException, IOException, XmlPullPar
...
-        
-        Profile profile = new Profile(version, authRef);
+
+        // If this is an old version of the profile file and is therefore missing uniqueId,
+        // get it from the update URL
+        boolean fromOld = false;
+        if (uniqueId == null) {
+            fromOld = true;
+            int startIndex = authRef.indexOf("download") + 9;
+            int endIndex = authRef.indexOf("profile") - 1;
```
Comment (orangejenny) [inline backend/src/org/commcare/xml/ProfileParser.java]: Paranoid, but perhaps this should use the indexOf(String str, int fromIndex) version and start from startIndex.

**Comment by orangejenny** [comment]
Looks fine, just minor comments.

File: backend/src/org/commcare/suite/model/Profile.java [hunk 3]
```diff
@@ -66,9 +69,12 @@ public Profile() {
      * is obsoleted by it. 
      * 
      */
-    public Profile(int version, String authRef) {
+    public Profile(int version, String authRef, String uniqueId, String displayName, boolean fromOld) {
         this.version = version;
         this.authRef = authRef;
+        this.uniqueId = uniqueId;
```
Comment (ctsims) [inline backend/src/org/commcare/suite/model/Profile.java]: these flags should be in the serialization, right?

**Comment by phillipm** [comment]
Looks good after you address @ctsims serialization comment.

File: backend/src/org/commcare/suite/model/Profile.java [hunk 4]
```diff
@@ -81,6 +92,30 @@ public int getID() {
...
+     * use the app name from Localization strings instead)
+     */
+	public String getDisplayName() {
+	    return this.displayName;
+	}
+
+    /**
+     * @return if this object was generated from an old version of the profile.ccpr file
+     */
+	public boolean oldVersion() {
```
Comment (ctsims) [inline backend/src/org/commcare/suite/model/Profile.java]: should be isOldVersion for conventions

File: tests/test/org/commcare/model/test/ProfileTests.java [hunk 5]
```diff
@@ -83,4 +110,23 @@ private Profile getBasicProfile() {
             throw PersistableSandbox.wrap("Error during profile test setup", e);
         }
     }
+
+    // Return a Profile object constructed from a profile.ccpr file explicitly containing the
+    // fields necessary for multiple apps support
+    private Profile getMultipleAppsProfile() {
```
Comment (ctsims) [inline tests/test/org/commcare/model/test/ProfileTests.java]: This should probably just be the same method as above and take in a string arg that is the path

**Pull request merged**